### PR TITLE
Week 3 walkthrough part 3&4: Json, post requests

### PR DESCRIPTION
**Pull request by charmychancey**
Create Json string on server. Parse Json string into JS object. Fetch content from server.

Part 4: add form and post features, maintain old code from json.

### CODE REVIEW

File: portfolio/src/main/webapp/script.js [hunk 1]
```diff
@@ -20,7 +20,22 @@ function off() {
...
-  const response = await fetch('/data');
-  const quote = await response.text();
-  document.getElementById('content-container').innerText = quote;
+    const response = await fetch('/data');
+    const hardCodedContent = await response.json();
+  
+    const contentListElement = document.getElementById("content-container");
+    contentListElement.innerHTML = '';
+    contentListElement.appendChild(
+        createListElement("Element 1: " + hardCodedContent[0]));
```
Comment (robotsaurus) [inline portfolio/src/main/webapp/script.js:29]: Element 0:

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -20,7 +20,22 @@ function off() {
...
+    contentListElement.appendChild(
+        createListElement("Element 1: " + hardCodedContent[1]));
+    contentListElement.appendChild(
+        createListElement("Element 1: " + hardCodedContent[2]));
```
Comment (robotsaurus) [inline portfolio/src/main/webapp/script.js]: Element 2:

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -20,7 +20,22 @@ function off() {
...
 async function getContent() {
-  const response = await fetch('/data');
-  const quote = await response.text();
-  document.getElementById('content-container').innerText = quote;
+    const response = await fetch('/data');
+    const hardCodedContent = await response.json();
+  
+    const contentListElement = document.getElementById("content-container");
+    contentListElement.innerHTML = '';
+    contentListElement.appendChild(
```
Comment (robotsaurus) [inline portfolio/src/main/webapp/script.js:28]: Do you need to append your `<li>..</li>` elements inside a ordered list (`<ol>`) or unordered list (`<ul>`) element?

Comment (charmychancey) [reply portfolio/src/main/webapp/script.js:28]: I used <ul> in index.html. That's what the walk-through used.

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (charmychancey) [reply portfolio/src/main/webapp/script.js]: typos also addressed in merge pull request #8 [e097599](https://github.com/charmychancey/step-my-portfolio/pull/6/commits/e097599814d93a03893c03d9b1699464f266e35c)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -23,10 +27,54 @@
...
+        if (sort) {
+            Arrays.sort(words);
+        }
+        comments.add(text);
+        comments.add(Arrays.toString(words));
+
+        // Respond with the result.
+        response.setContentType("text/html;");
+        response.getWriter().println(comments);
+        // response.sendRedirect("/index.html");
```
Comment (robotsaurus) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:69]: nit: remove dead code

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -20,7 +20,22 @@ function off() {
...
+        createListElement("Element 0: " + hardCodedContent[0]));
+    contentListElement.appendChild(
+        createListElement("Element 1: " + hardCodedContent[1]));
+    contentListElement.appendChild(
+        createListElement("Element 2: " + hardCodedContent[2]));
```
Comment (robotsaurus) [inline portfolio/src/main/webapp/script.js:33]: This is no longer hard coded content anymore as you updated this PR to also handle the post.  So you will need to update this to handle a variable number of comments.

File: portfolio/src/main/webapp/script.js (hunk 3 above)
Comment (charmychancey) [reply portfolio/src/main/webapp/script.js:28]: (It was just filler text to test how the server sent back the response)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -23,10 +27,54 @@
...
+
+        // Break the text into individual words.
+        String[] words = text.split("\\s*,\\s*");
+
+        // Sort the words.
+        if (sort) {
+            Arrays.sort(words);
+        }
+        comments.add(text);
+        comments.add(Arrays.toString(words));
```
Comment (robotsaurus) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:64]: This is going to add two entries to comments. From the code it seems like you only want the second one.  Am I reading this correctly?

Comment (charmychancey) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:64]: The end of the tutorial for part 3 didn't specify, so I wanted to keep both parts for now. The second one will most likely be removed in the next push.

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (charmychancey) [reply portfolio/src/main/webapp/script.js:33]: This part was not addressed in the part 4 tutorial. It seems like handling a variable set of data will go in hand with datastore for part 5 of walkthrough, so I will change that in the next pull request (part 5).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (charmychancey) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: oops, not dead code, should not have been commented out. Will fix.

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (robotsaurus) [reply portfolio/src/main/webapp/script.js]: I see the confusion.

Part 5 of the tutorial handles persistent storage.

For part 4 you still need to handle displaying a variable number of comments here.  The way you have it coded right now you are indexing out of the array when no comments are posted or when 1 comment is submitted.

Comment (charmychancey) [reply portfolio/src/main/webapp/script.js]: Oh I see, has been fixed!

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -20,7 +20,20 @@ function off() {
 }
 // asynchronously fetch content from server
 async function getContent() {
-  const response = await fetch('/data');
-  const quote = await response.text();
-  document.getElementById('content-container').innerText = quote;
+    const response = await fetch('/data');
```
Comment (robotsaurus) [inline portfolio/src/main/webapp/script.js:23]: Thanks for adding this.

nit: Can you verify the indentation here?

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -20,7 +20,20 @@ function off() {
...
+    const content = await response.json();
+  
+    const contentListElement = document.getElementById("content-container");
+    contentListElement.innerHTML = '';
+    for(let i=0; i < content.length; i++) {
```
Comment (robotsaurus) [inline portfolio/src/main/webapp/script.js:28]: nit: I think the style guide wants spaces around the equals sign.  so: let i = 0;

Comment (charmychancey) [reply portfolio/src/main/webapp/script.js]: oh i see, thanks!

**Pull request merged**